transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapChannel.java

```diff
@@ -0,0 +1,298 @@
+                    // Repeatedly read packets from the interface for as long as there are
+                    // packets to be read and the allocator handle tells us its okay...
+                    do {
+                        // Allocate an appropriately sized buffer in which to receive data.
+                        packetData = allocHandle.allocate(allocator);
+
+                        // Tell the allocator handle how much data we're attempting to read.
+                        allocHandle.attemptedBytesRead(packetData.writableBytes());
+
+                        int writerIndex = packetData.writerIndex();
+                        long readRes;
+
+                        // If the buffer has an accessible memory address, read packet data directly into the
+                        // buffer memory. Otherwise, get the internal NIO buffer associated with the ByteBuf
+                        // and read into that.
+                        if (packetData.hasMemoryAddress()) {
+                            readRes = tunTapFd.readTunTapPacketAddress(packetData.memoryAddress(), writerIndex,
+                                packetData.capacity());
+                        } else {
+                            ByteBuffer nioData = packetData.internalNioBuffer(writerIndex, packetData.writableBytes());
+                            readRes = tunTapFd.readTunTapPacket(nioData, nioData.position(), nioData.limit());
+                        }
+
+                        // Stop reading if no packet was received.
+                        if (readRes == 0) {
+                            allocHandle.lastBytesRead(-1);
+                            packetData.release();
+                            packetData = null;
+                            break;
+                        }
+
+                        // Extract the length of the packet and the protocol from the read result.
+                        int packetLen = (int) (readRes & 0xFFFFFFFF);
+                        int protocol = (int) ((readRes >> 32) & 0xFFFF);
+
+                        // Advance the writer index in the packet data buffer by the length of the packet.
+                        packetData.writerIndex(writerIndex + packetLen);
+
+                        // Allocate a TunTapPacket object containing the packet data and the protocol information.
+                        TunTapPacket packet = new TunTapPacket(protocol, packetData);
+                        _rcvdPackets.add(packet);
+
+                        // Avoid releasing the packet data buffer now that its wrapped in a TunTapPacket.
+                        packetData = null;
+
+                        // Tell the allocator handle about the newly received packet.
+                        allocHandle.incMessagesRead(1);
+                        allocHandle.lastBytesRead(packetLen);
+
+                    } while (allocHandle.continueReading());
+                } catch (Throwable ex) {
+                    if (packetData != null) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Change this condition so that it does not always evaluate to "false" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2583)

Reply:
This appears to be a bug in the checker.  Both readTunTapPacketAddress() and readTunTapPacket() can throw IOException.  If and when this happens, packetData will necessarily be non-null when the exception is thrown.  The catch() clause later in the method will catch the IOException, at which point the expression (packetData != null) will be true.

Reply:
agreed ... feel free to ignore.